AWS: check validity of KSM key before creating a new encrypted disk.

jsafrane opened the pull request:
AWS CreateVolume call does not check if referenced encryption key actually exists and returns a valid new AWS EBS volume even though an invalid key was specified. Later on it removes the EBS silently when its encryption fails.

To work around this buggy behavior we manually check that the key exists before calling CreateVolume.

Fixes #48438

/sig aws

Please review carefully. Can we safely assume that Kubernetes controller-manager can read encryption keys?

```release-note
aws: Kubernetes now checks existence of provided KSM (Key Management Service) key before creating an encrypted AWS EBS.
```

jsafrane commented:
/assign @justinsb 
/assign @gnufied

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1779,6 +1811,15 @@ func (c *Cloud) CreateDisk(volumeOptions *VolumeOptions) (KubernetesVolumeID, er
+	input := &kms.DescribeKeyInput{
+		KeyId: aws.String(keyId),
+	}
+	_, err := c.kms.DescribeKey(input)
```
So in theory at least this could fail for other reasons.  The most notable one is lack of permissions (and I'll touch on that separately).  Can we check the actual error code and wrap it, like e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L1457-L1463

I believe the error code should be "NotFoundException" (based on `aws --debug kms describe-key --key-id nope`)

It's probably a good idea to wrap the AWS error in both cases, as the raw AWS errors can be a bit cryptic.

i.e.

if code == "NotFoundException" {
return fmt.Errorf("KMS key %q not found: %v", keyId, err)
} else {
return fmt.Errorf("error querying for KMS key %q: %v", keyId, err)
}

justinsb commented:
Code looks great, but this is a potentially a new IAM permission, so we will need a release note I think.

(I'm actually not sure whether you need that IAM permission anyway to attach a volume with the KMS key - I think the answer is no).

There are two options I think:

1) If the error is "insufficient permissions" warn but carry on.  i.e. only error if we have permissions _and_ the key is not found.
2) Make it a requirement, as currently.

In kops, we scoped the KMS permissions just to the keys we know you need (e.g. for etcd volume encryption).  The unscoped DescribeKey permission feels pretty safe to me, but I also imagine that because it is KMS standards will be higher.  I'm inclined to do option 1 therefore.

jsafrane commented:
All I can do is to return either an error or created volume from a provisioner. The provisioner has no access to event recorder, so it can't put an event to the PVC "AccessDenied getting the key, but trying anyway".

I will block the provisioning only when NotFound is returned and just put a warning to log when AccessDenied or any other error is returned, hoping that CreateVolume fails gracefully.

jsafrane commented:
> I will block the provisioning only when NotFound is returned and just put a warning to log when AccessDenied or any other error is returned, hoping that CreateVolume fails gracefully.

Done

jsafrane commented:
> we will need a release note I think.

Added (please review / update)

gnufied commented:
/lgtm

cblecker commented:
/retest

cblecker commented:
/retest

cblecker commented:
/retest

jsafrane commented:
/assign @smarterclayton @thockin
for approvals

cblecker commented:
@jsafrane needs rebase and `hack/update-bazel.sh`

thockin commented:
Assign to me when LGTM'ed

jsafrane committed:
AWS: check validity of KSM key before creating a new encrypted disk.

AWS CreateVolume call does not check if referenced encryption key actually
exists and returns a valid new AWS EBS volume even though an invalid key
was specified. Later on it removes the EBS silently when its encryption fails.

To work around this buggy behavior we manually check that the key exists
before calling CreateVolume.

jsafrane commented:
rebased

jsafrane commented:
/assign @thockin 
please approve

thockin commented:
/lgtm
/approve

jsafrane commented:
/test all

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest
